### ISSUE TITLE
Try attach_file to improve IO performance, but did not get the desired effect

### ISSUE THREAD

**Issue by 97CBR** [StanfordLegion/legion#1463]
Recently I was reading [Integrating External Resources with a Task-Based Programming Model](https://legion.stanford.edu/pdfs/hipc2017.pdf). 
In this article I mentioned the ability of IRIS. After reading the article, I am interested in the fact that attach can produce a large IO performance improvement.

![image](https://user-images.githubusercontent.com/21171109/233838174-3c9634d9-5544-4921-85e0-7240d50ec1a4.png)

When looking through the legion code, I found that attach_file and attach_hdf5_file have been implemented into the legion.According to the description of the paper, I tried to use the C++ open() read() to open a file with a size of 1GB, just as described in the paper.. After reading the contents, I performed matrix operations as the baseline for IO. Then use attach_file to open a file of the same size to perform the same operation. It was found that attach_file did not improve performance as described in the paper.

I would like to ask, in what scenario can attach_file produce the IO performance improvement as shown in the paper, or can you provide similar sample code for reference, thank you very much.

**Comment by rohany** [comment]
Using attach does not make your IO any faster than normal. The benefits reported in that paper are around the overlap of IO with other parts of the application, which attach enables. The speed-ups there come from running the IO in the background while the application proceeds, versus a standard implementation that has to block until the IO completes before starting another iteration of the computation.

**Comment by 97CBR** [comment]
@rohany  I'm sorry, I didn't explain in the above question that the I/O in the above description  is the I/O after the overlap calculation.This means that the total execution time does not change whether or not attach_file is used.
Therefore, how to use attach_file to correctly overlap I/O ? Also, I wonder if attach_file must be used with CopyLauncher ?

**Comment by rohany** [comment]
Using attach and copies will allow for overlap to occur when there is computation or data movement to perform that does not depend on the write or read from disk. If your computation depends on the data from disk, then it cannot be overlapped. Does this answer your question?

**Comment by 97CBR** [comment]
Thank you very much. I understand what you mean. Please help me see if there are any problems with my experimental code.
Let me describe my experiment more clearly.
First, the experiment consists of loading data and computing matrices.
Load data section:
Example A: Use the open and read functions of C++ in a task to load a 1 GB binary data set that contains 2048 256 x 256 matrix data.
Example B: Use attach_file and copylauncher to load data of the same size.
Calculation part:
In examples A and B, IndexLauncher is used to start a computing task. In this example, IndexSpace and IndexPartition are used to split 16 partitions for IndexLauncher to perform computing.

Can the overlap effect be achieved in this scenario? If not, how was the experiment on attach_file written in the paper?

**Comment by rohany** [comment]
As you have described the experiment, it doesn't sound like there is too much overlap possible because the matrix multiplication depend on the data being loaded from disk. However, if you wrote your application like:

```
for i = 0, 2048:
  attach(data[i])
  matmul(data[i])
```
you would be able to start executing the matmuls that depend on each attached piece of data as soon as the data is read in, getting a little bit of overlap.

**Comment by z778520016** [comment]
But as the paper “Integrating_External_Resources_with_a_Task-Based_Programming_Model”，in  section “VI. EVALUATION A. Matrix Multiplication“  metioned：

**Comment by rohany** [comment]
I'll defer to @lightsighter or @jiazhihao for questions about specific benchmarks in that paper.

If you attach a file to the root region, then operations that read any partitions of that region depend on the attach completing. Examples like the one you linked attach a file to each sub-region in a partition, so any operations that depend only on a particular sub-region can proceed while other sub-regions are loaded in parallel.

**Comment by z778520016** [comment]
Since for example AttachLauncher.attach_hdf5 only support to read one file,I have tried:
`"data.h5 with two dataset(/matrix1,/matrix2)"`
`"using IndexCopyLauncher(Rect<1> color_bounds(0, 3)) to load data.h5 with one dataset(/matrix1)"`

Then both I tried to partition the root region to four sub-region to implement matrix multiplication,but the time shows it can not overlap automatically。

**Comment by rohany** [comment]
At this point I think I need to see some code to help you further. However, it doesn't sound like you're doing the correct thing -- remember that the partitions you are using to compute the matrix multiply should depend on the corresponding data you are loading from disk. So ideally you should have 4 matrix files, one corresponding to each sub-region of the matrix you are multiplying, so that once one sub-region is loaded, it can then be used by matrix multiplication tasks.

**Comment by z778520016** [comment]
Here is my code，but in this way，I think I just achieve the result of "in task I/O" with overlap，but failed to achieve the performance of “Iris” using attach.

How should I do to make full use of method mentioned in Iris paper such as “deferred execution” to get same performance in iris experiment。

By the way，I run my code in a single machine(OS:ubuntu 18.04)，not in a cluster。
 
```#include "legion.h"
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <fstream>but
#include <sys/syscall.h>
#include <thread>
#include <unistd.h>

using namespace std;
using namespace Legion;

enum TaskIDs{
    TOP_LEVEL_TASK_ID,
    CHECK_TASK_ID,
    INIT_TASK_ID,
};

enum FieldIDs{
    FID_X,
    FID_Y,
};

typedef FieldAccessor<READ_WRITE, double, 1, coord_t, Realm::AffineAccessor<double, 1, coord_t>> AccessorRW;
typedef FieldAccessor<READ_ONLY, double, 1, coord_t, Realm::AffineAccessor<double, 1, coord_t>> AccessorRO;

void top_level_task(const Task *task, const std::vector<PhysicalRegion> &regions, Context ctx, Runtime *runtime){
    struct timespec ts_start, ts_end;
    clock_gettime(CLOCK_MONOTONIC, &ts_start);
    printf("start time:%7.3f\n", 1.0 * ts_start.tv_sec + 1e-9 * ts_start.tv_nsec);
    int num_elements = 256 * 256 * 1024 * 2;
    int num_subregions = 1;

    Rect<1> ele_rect(0, num_elements - 1);
    IndexSpace is = runtime->create_index_space(ctx, ele_rect);
    FieldSpace input_fs = runtime->create_field_space(ctx);
    {
        FieldAllocator allocator = runtime->create_field_allocator(ctx, input_fs);
        allocator.allocate_field(sizeof(double), FID_X);
    }

    Rect<1> rect_B(0, 256 * 256 - 1);
    IndexSpace is_B = runtime->create_index_space(ctx, rect_B);
    FieldSpace fs_B = runtime->create_field_space(ctx);
    {
        FieldAllocator allocator = runtime->create_field_allocator(ctx, fs_B);
        allocator.allocate_field(sizeof(double), FID_X);
    }

    LogicalRegion input_lr = runtime->create_logical_region(ctx, is, input_fs);
    LogicalRegion input2_lr = runtime->create_logical_region(ctx, is, input_fs);
    LogicalRegion lr_B = runtime->create_logical_region(ctx, is, fs_B);
    runtime->fill_field<double>(ctx, lr_B, lr_B, FID_Y, 1.0);
    runtime->fill_field<double>(ctx, input_lr, input_lr, FID_X, 0.0);
    runtime->fill_field<double>(ctx, input2_lr, input2_lr, FID_X, 0.0);

    char a[] = "data_1D_1.h5";
    TaskLauncher load_launcher(INIT_TASK_ID, TaskArgument(a, sizeof(a)));
    load_launcher.add_region_requirement(RegionRequirement(input_lr, READ_WRITE, EXCLUSIVE, input_lr));
    load_launcher.region_requirements[0].add_field(FID_X);
    runtime->execute_task(ctx, load_launcher);

    char b[] = "data_1D_2.h5";
    TaskLauncher load_launcher2(INIT_TASK_ID, TaskArgument(a, sizeof(a)));
    load_launcher2.add_region_requirement(RegionRequirement(input2_lr, READ_WRITE, EXCLUSIVE, input2_lr));
    load_launcher2.region_requirements[0].add_field(FID_X);
    runtime->execute_task(ctx, load_launcher2);

    Rect<1> color_bounds(0, num_subregions - 1);
    IndexSpace colo_is = runtime->create_index_space(ctx, color_bounds);
    IndexPartition ip = runtime->create_equal_partition(ctx, is, colo_is);
    LogicalPartition input_lp = runtime->get_logical_partition(ctx, input_lr, ip);
    LogicalPartition input2_lp = runtime->get_logical_partition(ctx, input2_lr, ip);

    ArgumentMap arg_map;

    int task_num = 1;
    IndexLauncher check_launcher(CHECK_TASK_ID, colo_is, TaskArgument(&task_num, sizeof(int)), arg_map);
    check_launcher.add_region_requirement(RegionRequirement(input_lp, 0, READ_ONLY, EXCLUSIVE, input_lr));
    check_launcher.region_requirements[0].add_field(FID_X);
    check_launcher.add_region_requirement(RegionRequirement(lr_B, READ_ONLY, EXCLUSIVE, lr_B));
    check_launcher.region_requirements[1].add_field(FID_Y);
    FutureMap future1 = runtime->execute_index_space(ctx, check_launcher);

    int task_num = 2;
    IndexLauncher check2_launcher(CHECK_TASK_ID, colo_is, TaskArgument(&task_num, sizeof(int)), arg_map);
    check2_launcher.add_region_requirement(RegionRequirement(input2_lp, 0, READ_ONLY, EXCLUSIVE, input2_lr));
    check2_launcher.region_requirements[0].add_field(FID_X);
    check2_launcher.add_region_requirement(RegionRequirement(lr_B, READ_ONLY, EXCLUSIVE, lr_B));
    check2_launcher.region_requirements[1].add_field(FID_Y);
    FutureMap future2 = runtime->execute_index_space(ctx, check2_launcher);

    future1.wait_all_results();
    future2.wait_all_results();
    clock_gettime(CLOCK_MONOTONIC, &ts_end);
    printf("end time:%7.3f\n", 1.0 * ts_end.tv_sec + 1e-9 * ts_end.tv_nsec);
    double all_time = ((1.0 * (ts_end.tv_sec - ts_start.tv_sec)) + (1e-9 * (ts_end.tv_nsec - ts_start.tv_nsec)));
    printf("ALL RUN TIME = %7.3fs\n", all_time);

    runtime->destroy_logical_region(ctx, input_lr);
    runtime->destroy_logical_region(ctx, input2_lr);
    runtime->destroy_logical_region(ctx, lr_B);
    runtime->destroy_field_space(ctx, input_fs);
    runtime->destroy_field_space(ctx, fs_B);
    runtime->destroy_index_space(ctx, is);
    runtime->destroy_index_space(ctx, is_B);
    runtime->destroy_index_space(ctx, color_is);
}

void check_task(const Task *task, const std::vector<PhysicalRegion> &regions, Context ctx, Runtime *runtime){
    const int task_num = *(const int *)task->args;
    struct timespec start_calcu, end_calcu;
    clock_gettime(CLOCK_MONOTONIC, &start_calcu);
    AccessorRO acc_x(regions[0], FID_X);
    AccessorRO acc_z(regions[1], FID_Y);
    Rect<1> rect = runtime->get_index_space_domain(ctx, task->regions[0].region.get_index_space());
    const int readSize = rect.hi.x - rect.lo.x + 1;
    printf("start calcu %d,time:%7.3f\n", task_num, 1.0 * start_calcu.tv_sec + 1e-9 * start_calcu.tv_nsec);

    double* sum = new double[1];
    PointInRectIterator<1> pir(rect);
    int matrixSize = 256;
    int matrixSizes = matrixSize * matrixSize;
    int matrixCnt = readSize / matrixSizes;
    double* result = new double[matrixSizes * matrixCnt];
    for(int cnt = 0; i < matrixCnt; ++cnt){
        for(int i = 0; i < matrixSize; ++i){
            int indexL = cnt * matrixSizes + i * matrixSize;
            for(int j = 0; j < matrixSize; ++j){
                sum[0] = 0;
                for(int k = 0; k < matrixSize; k++){
                    sum[0] +=acc_x[*pir + indexL + k] * acc_z[*pir + k * matrixSize + j];
                }
                result[indexL + j] = sum[0];
            }
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &end_calcu);
    printf("end calcu %d,time:%7.3f\n", task_num, 1.0 * end_calcu.tv_sec + 1e-9 * end_calcu.tv_nsec);
    delete [] num;
    delete [] result;
}

void init_task(const Task *task, const std::vector<PhysicalRegion> &regions, Context ctx, Runtime *runtime){
    struct timespec start_load, end_load;
    const char* filename = (const char *)task->args;
    int num_elements = 256 * 256 * 2048;
    Rect<1> elem_rect(0, num_elements - 1);
    IndexSpace is = runtime->create_index_space(ctx, num_elements);
    FieldSpace input_fs = runtime->create_field_space(ctx);
    {
        FieldAllocator allocator = runtime->create_field_allocator(ctx, input_fs);
        allocator.allocate_field(sizeof(double), FID_X);
    }
    LogicalRegion cp_lr = runtime->create_logical_region(ctx, is, input_fs);

    clock_gettime(CLOCK_MONOTONIC, &start_load);
    printf("start load %s,time:%7.3f\n", filename, 1.0 * start_load.tv_sec + 1e-9 * start_load.tv_nsec);
    PhysicalRegion cp_pr;
    std::map<FieldID, const char*> field_map;
    field_map[FID_X] = "matrix";
    AttachLauncher al(LEGION_EXTERNAL_HDF5_FILE, cp_lr, cp_lr);
    al.attach_hdf5(filename, field_map, LEGION_FILE_READ_WRITE);
    cp_pr = runtime->attach_external_resource(ctx, al);

    CopyLauncher copy_launcher;
    copy_launcher.add_copy_requirements(
        RegionRequirement(cp_lr, READ_ONLY, EXCLUSIVE, cp_lr),
        RegionRequirement(regions[0].get_logical_region(), READ_WRITE, EXCLUSIVE, regions[0].get_logical_region());
    );
    copy_launcher.add_src_field(0, FID_X);
    copy_launcher.add_dst_field(0, FID_X);
    runtime->issue_copy_operation(ctx, copy_launcher);
    clock_gettime(CLOCK_MONOTONIC, &end_load);
    printf("end load %s,time:%7.3f\n", filename, 1.0 * end_load.tv_sec + 1e-9 * end_load.tv_nsec);
}

int main(int argc, char **argv){
    Runtime::set_top_level_task_id(TOP_LEVEL_TASK_ID);
    {
        TaskVariantRegistrar registrar(TOP_LEVEL_TASK_ID, "top_level");
        registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
        Runtime::preregistrar_task_variant<top_level_task>(registrar, "top_level");
    }
    {
        TaskVariantRegistrar registrar(CHECK_TASK_ID, "check");
        registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
        registrar.set_leaf();
        Runtime::preregistrar_task_variant<check_task>(registrar, "check");
    }
    {
        TaskVariantRegistrar registrar(INIT_TASK_ID, "init");
        registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
        Runtime::preregistrar_task_variant<init_task>(registrar, "init");
    }
}

> Recently I was reading [Integrating External Resources with a Task-Based Programming Model](https://legion.stanford.edu/pdfs/hipc2017.pdf). In this article I mentioned the ability of IRIS. After reading the article, I am interested in the fact that attach can produce a large IO performance improvement.
> 
> ![image](https://user-images.githubusercontent.com/21171109/233838174-3c9634d9-5544-4921-85e0-7240d50ec1a4.png)
> 
> When looking through the legion code, I found that attach_file and attach_hdf5_file have been implemented into the legion.According to the description of the paper, I tried to use the C++ open() read() to open a file with a size of 1GB, just as described in the paper.. After reading the contents, I performed matrix operations as the baseline for IO. Then use attach_file to open a file of the same size to perform the same operation. It was found that attach_file did not improve performance as described in the paper.
> 
> I would like to ask, in what scenario can attach_file produce the IO performance improvement as shown in the paper, or can you provide similar sample code for reference, thank you very much.

**Comment by rohany** [comment]
So your load task takes the entire root regions `input_lr` and `input_lr2` as read-write (and attaches to them I presume). So anything that uses `input_lr` or children of them, like the partitions you create after depend on the load completing, so no pipelining is possible here.

**Comment by z778520016** [comment]
But the I/O performance I realized in the above code much lower than that in "Iris" (mentioned in first comment`s diagram).And
I am curioused about how to achieve the same performance with attach.

**Comment by rohany** [comment]
I don't know how that particular benchmark was written -- maybe @lightsighter remembers.

**Comment by lightsighter** [comment]
To be clear, using attach/detach isn't going to do anything to speed-up performance over carefully handwritten file I/O benchmarks. It will just make it easier to do the overlapping presuming you have enough task parallelism. A cursory overview of the code above suggest that there is not much task parallelism at all. If you don't have any task parallelism to hide the latency of file I/O then you're going to see the full latency of that I/O. That is not a surprising result.

I'll also note that looking at graphs is almost useless. We provide a profiler for Legion so that you can make profiles and see where time is being spent. Speculating without a profile to look at is a waste of time.

**Comment by z778520016** [comment]
Thanks for lightsighter`s and rohany`s help,and i will do more experiment with enough task parallelism and using the profiler。

**Comment by z778520016** [comment]
@rohany @lightsighter We are interested in the first experiment(Matrix Multiplication) of the paper “Integrating External Resources with a Task-Based Programming Model”, but failed to reproduce it, could you provide the source code for the experiment, or code about how to implement a good overlapping。Thank you very much。
